hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java

```diff
@@ -209,10 +210,18 @@ public static byte[] avroToJson(GenericRecord record, boolean pretty) throws IOE
   private static ByteArrayOutputStream avroToJsonHelper(GenericRecord record, boolean pretty) throws IOException {
     DatumWriter<Object> writer = new GenericDatumWriter<>(record.getSchema());
     ByteArrayOutputStream out = new ByteArrayOutputStream();
-    JsonEncoder jsonEncoder = EncoderFactory.get().jsonEncoder(record.getSchema(), out, pretty);
-    writer.write(record, jsonEncoder);
-    jsonEncoder.flush();
-    return out;
+    try {
+      JsonEncoder jsonEncoder = EncoderFactory.get().jsonEncoder(record.getSchema(), out, pretty);
+      writer.write(record, jsonEncoder);
+      jsonEncoder.flush();
+      return out;
+    } catch (ClassCastException | NullPointerException ex) {
+      // NullPointerException will be thrown in cases where the field values are missing
+      // ClassCastException will be thrown in cases where the field values do not match the schema type
+      // Fallback to using `toString` which also returns json but without a pretty-print option
+      out.write(record.toString().getBytes(StandardCharsets.UTF_8));
```

Comment:
What do we get for `record.toString` when `NullPointerException` is thrown?

Reply:
Hmm, seems like a `null` constant for empty field.

Reply:
So when the transformed JSON string got converted back into avro, the schema could change right?

Reply:
You get a string that represents the json of the object, it does not do any validation on types/nullability. See the tests that are added for a sample.

Reply:
> So when the transformed JSON string got converted back into avro, the schema could change right?

The case here is when you have some data and are trying to convert it to avro and it fails. https://github.com/apache/hudi/blob/master/hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/HoodieStreamerUtils.java#L164

Reply:
> You get a string that represents the json of the object, it does not do any validation on types/nullability

I kind of remember we have some cases for converting the json into avro then back to json again operations for our commit metadata.

Reply:
Ok, I'm not familiar with that flow. If this is breaking that flow, I can just make a new method

Reply:
If the schema does not really change for that, it is okay, maybe we can add some use cases.

Reply:
One concern I have is that this could hide some exception and then we don't catch something in our initial testing for some more critical timeline related flow. I think I've convinced myself there should just be a new method like "safeToJson" that does not throw an exception that we use in the error table/writer cases since those are not as critical to Hudi.

Reply:
>  think I've convinced myself there should just be a new method like "safeToJson" that does not throw an exception that we use in the error table/writer cases since those are not as critical to Hudi.

+1

Reply:
I've pushed an update